### ISSUE TITLE
2D Slide Label Implementation Discussion

### ISSUE THREAD

**Issue by SidHarder** [openaplis/ap-pm#1]
@dspencer-mawd :  [This is the document](https://github.com/openaplis/ap-pm/blob/master/TwoDBarcodeProjectScope.md) that I created to define the first project for us to work on.  The scope of this project is pretty small shouldn't take to long.  

Before I start figuring out the steps to accomplish it I want you to go in and add/subtract what ever I've gotten wrong.

You can make comments hear or edit the document directly.

**Comment by SidHarder** [comment]
@dspencer-mawd : 

`Keep current slide printing process as is, however, include scannable linear barcode on slides (I am 
close to being able to implement this).`

Why wouldn't we print a 2D barcode instead of a linear barcode.  Linear barcodes will limit us in a big way as far as how many characters the ID can contain.


@SidHarder 

I can natively print a linear barcode out of CoPath.  For a 2D barcode, I cannot do it directly out of CoPath.  To get the 2D barcode, I believe we would need to interface with our slide printers as opposed to printing to them directly out of CoPath as we do now.  The end game is definitely a 2D barcode, but if I can encode enough info in a linear barcode that uniquely identifies the individual stain in CoPath, and have that rolled out within 2 days, we could start tracking without any process changes in printing slides.

**Comment by SidHarder** [comment]
@dspencer-mawd : 

`My plan is to use this id and strip the prefix.`

An example of one of our ID's for a Slide is: HSLD18-1234.3A.2.  The first four digits of the id represent what the material being scanned is.  Because we know unambiguously what the material being scanned is we can perform different actions is in different contexts and we can save steps for the user.  This is a critical point.

I'm thinking that we generate an ID that is stored in MYSQL and is associated to the keys extracted from CoPath.  This does make the ID a bit longer and thus requires the use of a 2D barcode.

**Comment by SidHarder** [comment]
@dspencer-mawd : 

`Linear barcode will contain the following: specimen_id, part_inst, stain_inst, level, this information is enough to uniquely identify every stain that is tracked within CoPathPlus, as well as the block it is tied to.`

Can you post an example of what these ID's look like: specimen_id, part_inst, stain_inst, level.

**Comment by dspencer-kc** [comment]
Copath uses an internal ID number as the primary key that does not show anywhere, and then there is a formatted number as well.  The "D" specimen class signifies it is a derm case.  If needed, we could also lookup the "part" information as well which would give use the material and what site it came from.


specimen_id | specclass_id | specnum_formatted | numwheel_id | specnum_year | specnum_num
-- | -- | -- | -- | -- | --
mwd0000402 | D | D18-1838 | D | 2018 | 1838


Then the stain/slide is identified with the following info:


specimen_id | part_inst | stain_inst | stainprocess_id | order_date | orderedby_id | slidecount | stainstatus_id | status_date | label_text | stainproclog_id | log_comment | requestclass_id | block_inst | ppath_link | wkdept_id | unstained_available | sequence
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
mwd0000402 | 1 | 1 | $hinit | 30-Jan-18 |   | 1 | $ord | 30-Jan-18 | H&E | cox1 |   | co3 | 1 |   | mwd4 |   | 1

**Comment by SidHarder** [comment]
`D18-1838`

Is this the Accession No?

**Comment by dspencer-kc** [comment]
Yes - this is our Accession Number / Case Number

**Comment by SidHarder** [comment]
@dspencer-mawd : So if you have 5 specimens for one encounter, does each specimen get it's own accession number?

**Comment by dspencer-kc** [comment]
@SidHarder No, each accession has multiple "parts" or specimens.  So it would be one accession number with 5 parts tied to it.

**Comment by SidHarder** [comment]
@dspencer-mawd could you send me the make and model of you direct to slide printers.  And do you have one that is not in use that we can play with.

**Comment by dspencer-kc** [comment]
@SidHarder I have 5 or 6 that are not in use that you can play with.  Also - they are USB, they offer a a USB over Network adapter to make the network compatible.  They also have software included. 

HistoCore PERMA S by Leica.  I have one set up directly to my computer.  If it would help I could set up a spare computer with one attached and give you remote access.

**Comment by SidHarder** [comment]
@dspencer-mawd I need you to confirm something for me.  The preferred way to print to these printers would be to use the PT Lab software that comes with it.  This would simplify creation of the label itself as this software allows you to easily create templates.  I expected to find a way for this software to monitor a folder where I could drop in formatted text files that it would pickup and print.  As I looked through the documentation I didn't see any options for this type of use.

Is it possible that they have some add on feature that would help accomplish this.

Could you confirm for me that I'm not missing something?

**Comment by dspencer-kc** [comment]
@SidHarder My understanding was that it could watch a folder, but we never implemented that.  I will confirm, I reached out to one of the reps that sold us the slide printers and hope to hear back soon.  I have the software on a USB drive, and installed on my PC but do not see a way to watch a folder.  I'll find out.

**Comment by dspencer-kc** [comment]
@SidHarder After working with the software some, and speaking briefly to our sales rep, it appears this version does not have a folder watch.  The only integration I see is the scanning of a 2d barcode from a block to populate the template - although this does not make sense to me, as when the block is printed, the stains tied to that block are not going to be set.  The only information that could be included on the barcode would be patient information and accession / block id.  I am looking closer in to this, and will give you an update when I can find more out.

**Comment by SidHarder** [comment]
@dspencer-mawd if we are unable to use the PT Lab software, plan B looks like this:
1. Use bwip-js to generate a 2D barcode
2. Use jimp to generate a PNG file with the 2D barcode on it
3. Implement a Print Monitor to run on the Windows PC connected to the printer to poll a folder and print PNG files that show up.

Using bwip-js and jimp are both trivial to setup and use.  I can use code from an existing print monitor that we use here for step 3.

So we build a page that shows the users the slides that have been ordered in CoPath.  They hit the print button and we loop through each slide generating an ID that is appropriate and storing that ID along with its CoPath ID's in MYSQL.  We then send that ID along with other data to a service that will generate the slide label image.  The service shoots the slide label image over to a folder which the Print Monitor then picks up and the slide prints out.

**Comment by dspencer-kc** [comment]
@SidHarder I'm waiting for a Leica software/interface engineer to call, should have some more info after that discussion.  Plan B should work as well.

On a related note, I am working with Cerner and General Data to interface our 3 cassette engravers with CoPath.  The General Data /Pulsar software just watches a folder and then prints the flat files dropped there.  Cerner wants $10,000 to enable each interface folder, so I will most likely have one interface that generates the flatfile, and then run those files through mirth, and distribute them to the 3 folders watched by the corresponding engravers.  By going through Mirth, I could also write to MYSQL when cassettes are printed.  This could be useful to start tracking at the block level, thought I would mention it.  If we have a problem querying the CoPath database for stains, we could limit that query to include just the blocks that have been printed.

**Comment by dspencer-kc** [comment]
@SidHarder Also - I think we should implement this in a way so that slides are printed by scanning blocks - like you were pushing for. We may have to do it in batches until we can have slide printers at every workstation, but real-estate wise that is not practical until we are in our new space.

**Comment by dspencer-kc** [comment]
@SidHarder Good news on the software that works with our Slide Printers.  There is a piece of software that should have been included but was not.  Nice Label is used, this is what facilitates a folder being watched, and also has a manager of what fields are mapped where based on the flat file that is being dropped.  Our rep should have 11 copies of the software and licenses soon.  He also mentioned these have the same interface capability as Leica IPS.

**Comment by dspencer-kc** [comment]
@SidHarder There are two piece of software I believe, Nicelabel, and Nicewatch.  I think these should be pretty straightforward to configure.

**Comment by SidHarder** [comment]
@dspencer-mawd that is awesome news.

**Comment by dspencer-kc** [comment]
@SidHarder OK, another update, the vendor has placed an order for the software and it has to be shipped to us, and their estimate for that is two weeks.  In the meantime, I will focus on our cassette engraver interface including the Mirth portion that will write to the mysql database.

**Comment by dspencer-kc** [comment]
@SidHarder Do you have a recommended BarcodeScanner for reading 2-d barcodes on blocks?

**Comment by dspencer-kc** [comment]
@SidHarder Have you used Vue to build a page to interface with the barcode scanner?  If so - if you have any recommended libraries / examples that would be helpful.

**Comment by SidHarder** [comment]
@dspencer-mawd  this is one option: https://github.com/billhsu/jUART

**Comment by dspencer-kc** [comment]
> @dspencer-mawd :
> 
> `My plan is to use this id and strip the prefix.`
> 
> An example of one of our ID's for a Slide is: HSLD18-1234.3A.2. The first four digits of the id represent what the material being scanned is. Because we know unambiguously what the material being scanned is we can perform different actions is in different contexts and we can save steps for the user. This is a critical point.
> 
> I'm thinking that we generate an ID that is stored in MYSQL and is associated to the keys extracted from CoPath. This does make the ID a bit longer and thus requires the use of a 2D barcode.

@SidHarder  - I am creating the ID that will be generated to identify the slides.  You gave the example "HSLD18-1234.3A.2" and mentioned the first 4 digits represent material type.  Do you have a set of predifined values to use for this?  This ID will be generated, and then the copath identifiers will be stored as well.

**Comment by SidHarder** [comment]
These are the prefixes we currently use

HSLD - histology slide
HBLK - histology block
YREQ - ypi requisition
SBDG - security badge
CTNR - specimen container
DCMT - document
CLNT - client
